File: apple/TriageCompanion/JiraClient.swift
import Foundation

struct JiraClient {
    private let settings: JiraSettings
    private let http: HTTPClient

    init(baseURL: String, email: String, apiToken: String, cloudID: String?, http: HTTPClient = HTTPClient()) throws {
        self.settings = JiraSettings(
            baseURL: try normalizedJiraBaseURL(baseURL),
            email: try cleanRequiredText(email, label: "Jira email"),
            apiToken: try cleanRequiredText(apiToken, label: "Jira API token"),
            cloudID: cloudID
        )
        self.http = http
    }

    func listOpenTickets() async throws -> [JiraTicketItem] {
        var tickets: [JiraTicketItem] = []
        var nextPageToken: String?
        var seenTokens: Set<String> = []
        var startAt = 0
        let pageSize = 100

        while true {
            var queryItems = [
                URLQueryItem(name: "jql", value: "assignee = currentUser() AND resolution = Unresolved ORDER BY updated DESC"),
                URLQueryItem(name: "fields", value: "summary,status,priority,issuetype,reporter,updated,resolution"),
                URLQueryItem(name: "maxResults", value: String(pageSize))
            ]
            if settings.apiKind == .dataCenter {
                queryItems.append(URLQueryItem(name: "startAt", value: String(startAt)))
            } else if let nextPageToken {
                queryItems.append(URLQueryItem(name: "nextPageToken", value: nextPageToken))
            }

            let url = urlWithQuery(try settings.routeURL(path: settings.searchPath), queryItems: queryItems)
            let (payload, _) = try await http.jsonObject(
                from: request(url: url, headers: settings.headers),
                serviceName: "Jira"
            )
            guard let issueRecords = payload["issues"] as? [[String: Any]] else {
                throw TriageCompanionError.message("Jira search response must include an issues array.")
            }
            if issueRecords.count > pageSize {
                throw TriageCompanionError.message("Jira search response exceeded the requested page size.")
            }
            for issue in issueRecords {
                tickets.append(try ticketItem(issue))
            }

            if settings.apiKind == .dataCenter {
                startAt += issueRecords.count
                let total = payload["total"] as? Int
                if issueRecords.isEmpty || issueRecords.count < pageSize || (total != nil && startAt >= total!) {
                    break
                }
            } else {
                guard let token = payload["nextPageToken"] as? String else {
                    break
                }
                let cleanToken = try cleanRequiredText(token, label: "Jira nextPageToken")
                if issueRecords.isEmpty {
                    throw TriageCompanionError.message("Jira search returned an empty page before pagination finished.")
                }
                if !seenTokens.insert(cleanToken).inserted {
                    throw TriageCompanionError.message("Jira search pagination repeated a page token.")
                }
                nextPageToken = cleanToken
            }
        }

        return tickets.sorted { $0.updatedAt > $1.updatedAt }
    }

    private func ticketItem(_ issue: [String: Any]) throws -> JiraTicketItem {
        let key = try issueKey(string(issue, key: "key", context: "Jira issue"))
        let fields = try object(issue, key: "fields", context: "Jira issue \(key)")
        if fields["resolution"] != nil, !(fields["resolution"] is NSNull) {
            throw TriageCompanionError.message("Jira issue \(key) was resolved.")
        }

        let issueType = try string(try object(fields, key: "issuetype", context: "Jira issue \(key)"), key: "name", context: "Jira issue \(key) type")
        let status = try string(try object(fields, key: "status", context: "Jira issue \(key)"), key: "name", context: "Jira issue \(key) status")
        let summary = try string(fields, key: "summary", context: "Jira issue \(key)")
        let updatedAt = try parseServiceDate(
            string(fields, key: "updated", context: "Jira issue \(key)"),
            label: "Jira issue updated"
        )
        let priority = (fields["priority"] as? [String: Any]).flatMap { try? string($0, key: "name", context: "Jira issue \(key) priority") }
        let reporterObject = fields["reporter"] as? [String: Any]
        let reporter = reporterObject.flatMap { reporter in
            (try? string(reporter, key: "displayName", context: "Jira issue \(key) reporter"))
                ?? (try? string(reporter, key: "emailAddress", context: "Jira issue \(key) reporter"))
        }
        guard let url = URL(string: "\(settings.baseURL)/browse/\(encodedPathComponent(key))") else {
            throw TriageCompanionError.message("Could not build Jira issue URL.")
        }

        return JiraTicketItem(
            id: key,
            key: key,
            issueType: issueType,
            status: status,
            priority: priority,
            reporter: reporter,
            updatedAt: updatedAt,
            summary: summary,
            url: url
        )
    }

    private func object(_ value: [String: Any], key: String, context: String) throws -> [String: Any] {
        guard let object = value[key] as? [String: Any] else {
            throw TriageCompanionError.message("\(context) must include \(key).")
        }

        return object
    }

    private func string(_ value: [String: Any], key: String, context: String) throws -> String {
        guard let stringValue = value[key] as? String else {
            throw TriageCompanionError.message("\(context) must include \(key).")
        }

        return try cleanRequiredText(stringValue, label: "\(context) \(key)")
    }

    private func issueKey(_ value: String) throws -> String {
        guard value.range(of: #"^[A-Za-z][A-Za-z0-9_]*-\d+$"#, options: .regularExpression) != nil else {
            throw TriageCompanionError.message("Jira issue key must use project-key-number format.")
        }

        return value.uppercased()
    }
}

private struct JiraSettings {
    let baseURL: String
    let email: String
    let apiToken: String
    let cloudID: String?

    var apiKind: JiraAPIKind {
        if cloudID != nil {
            return .cloud
        }
        guard let host = URL(string: baseURL)?.host?.lowercased() else {
            return .cloud
        }

        return host.hasSuffix(".atlassian.net") ? .cloud : .dataCenter
    }

    var searchPath: String {
        apiKind == .dataCenter ? "/rest/api/2/search" : "/rest/api/3/search/jql"
    }

    var headers: [String: String] {
        [
            "Authorization": authorizationHeader,
            "Accept": "application/json",
            "User-Agent": "triage-companion"
        ]
    }

    private var authorizationHeader: String {
        if apiKind == .dataCenter {
            return "Bearer \(apiToken)"
        }

        let encoded = Data("\(email):\(apiToken)".utf8).base64EncodedString()
        return "Basic \(encoded)"
    }

    func routeURL(path: String) throws -> URL {
        let apiBase: String
        if let cloudID {
            apiBase = "https://api.atlassian.com/ex/jira/\(encodedPathComponent(cloudID))"
        } else {
            apiBase = baseURL
        }
        guard let url = URL(string: "\(apiBase)\(path)") else {
            throw TriageCompanionError.message("Jira API route was invalid.")
        }

        return url
    }
}

private enum JiraAPIKind {
    case cloud
    case dataCenter
}

File: android/app/src/main/kotlin/com/triagecompanion/app/JiraClient.kt
package com.triagecompanion.app

import android.util.Base64
import java.net.URI
import java.net.URISyntaxException
import org.json.JSONObject

class JiraClient(
    baseURL: String,
    email: String,
    apiToken: String,
    cloudID: String?,
    private val http: HttpClient = HttpClient(),
) {
    private val settings = JiraSettings(
        baseURL = normalizedJiraBaseURL(baseURL),
        email = cleanRequiredText(email, label = "Jira email"),
        apiToken = cleanRequiredText(apiToken, label = "Jira API token"),
        cloudID = cloudID,
    )

    suspend fun listOpenTickets(): List<JiraTicketItem> {
        val tickets = mutableListOf<JiraTicketItem>()
        var nextPageToken: String? = null
        val seenTokens = mutableSetOf<String>()
        var startAt = 0
        val pageSize = 100

        while (true) {
            val queryItems = mutableListOf(
                "jql" to "assignee = currentUser() AND resolution = Unresolved ORDER BY updated DESC",
                "fields" to "summary,status,priority,issuetype,reporter,updated,resolution",
                "maxResults" to pageSize.toString(),
            )
            if (settings.apiKind == JiraAPIKind.DATA_CENTER) {
                queryItems.add("startAt" to startAt.toString())
            } else if (nextPageToken != null) {
                queryItems.add("nextPageToken" to nextPageToken)
            }

            val url = urlWithQuery(settings.routeURL(settings.searchPath), queryItems)
            val (payload, _) = http.jsonObject(url, settings.headers, serviceName = "Jira")
            val issueRecords = payload.optJSONArray("issues")
                ?: throw TriageCompanionException("Jira search response must include an issues array.")
            if (issueRecords.length() > pageSize) {
                throw TriageCompanionException("Jira search response exceeded the requested page size.")
            }
            for (index in 0 until issueRecords.length()) {
                val issue = issueRecords.opt(index) as? JSONObject
                    ?: throw TriageCompanionException("Jira search response must include an issues array.")
                tickets.add(ticketItem(issue))
            }

            if (settings.apiKind == JiraAPIKind.DATA_CENTER) {
                startAt += issueRecords.length()
                val total = payload.opt("total") as? Int
                if (issueRecords.length() == 0 || issueRecords.length() < pageSize ||
                    (total != null && startAt >= total)
                ) {
                    break
                }
            } else {
                val token = payload.opt("nextPageToken") as? String ?: break
                val cleanToken = cleanRequiredText(token, label = "Jira nextPageToken")
                if (issueRecords.length() == 0) {
                    throw TriageCompanionException("Jira search returned an empty page before pagination finished.")
                }
                if (!seenTokens.add(cleanToken)) {
                    throw TriageCompanionException("Jira search pagination repeated a page token.")
                }
                nextPageToken = cleanToken
            }
        }

        return tickets.sortedByDescending { it.updatedAt }
    }

    private fun ticketItem(issue: JSONObject): JiraTicketItem {
        val key = issueKey(string(issue, "key", context = "Jira issue"))
        val fields = obj(issue, "fields", context = "Jira issue $key")
        if (fields.has("resolution") && !fields.isNull("resolution")) {
            throw TriageCompanionException("Jira issue $key was resolved.")
        }

        val issueType = string(obj(fields, "issuetype", "Jira issue $key"), "name", "Jira issue $key type")
        val status = string(obj(fields, "status", "Jira issue $key"), "name", "Jira issue $key status")
        val summary = string(fields, "summary", context = "Jira issue $key")
        val updatedAt = parseServiceDate(
            string(fields, "updated", context = "Jira issue $key"),
            label = "Jira issue updated",
        )
        val priority = fields.optJSONObject("priority")?.let { priorityObject ->
            try {
                string(priorityObject, "name", "Jira issue $key priority")
            } catch (_: TriageCompanionException) {
                null
            }
        }
        val reporter = fields.optJSONObject("reporter")?.let { reporterObject ->
            try {
                string(reporterObject, "displayName", "Jira issue $key reporter")
            } catch (_: TriageCompanionException) {
                try {
                    string(reporterObject, "emailAddress", "Jira issue $key reporter")
                } catch (_: TriageCompanionException) {
                    null
                }
            }
        }

        return JiraTicketItem(
            id = key,
            key = key,
            issueType = issueType,
            status = status,
            priority = priority,
            reporter = reporter,
            updatedAt = updatedAt,
            summary = summary,
            url = "${settings.baseURL}/browse/${encodedPathComponent(key)}",
        )
    }

    private fun obj(value: JSONObject, key: String, context: String): JSONObject =
        value.optJSONObject(key)
            ?: throw TriageCompanionException("$context must include $key.")

    private fun string(value: JSONObject, key: String, context: String): String {
        val stringValue = value.opt(key) as? String
            ?: throw TriageCompanionException("$context must include $key.")

        return cleanRequiredText(stringValue, label = "$context $key")
    }

    private fun issueKey(value: String): String {
        if (!value.matches(Regex("^[A-Za-z][A-Za-z0-9_]*-\\d+$"))) {
            throw TriageCompanionException("Jira issue key must use project-key-number format.")
        }

        return value.uppercase()
    }
}

private enum class JiraAPIKind {
    CLOUD,
    DATA_CENTER,
}

private class JiraSettings(
    val baseURL: String,
    val email: String,
    val apiToken: String,
    val cloudID: String?,
) {
    val apiKind: JiraAPIKind
        get() {
            if (cloudID != null) {
                return JiraAPIKind.CLOUD
            }
            val host = try {
                URI(baseURL).host?.lowercase()
            } catch (_: URISyntaxException) {
                null
            } ?: return JiraAPIKind.CLOUD

            return if (host.endsWith(".atlassian.net")) JiraAPIKind.CLOUD else JiraAPIKind.DATA_CENTER
        }

    val searchPath: String
        get() = if (apiKind == JiraAPIKind.DATA_CENTER) "/rest/api/2/search" else "/rest/api/3/search/jql"

    val headers: Map<String, String>
        get() = mapOf(
            "Authorization" to authorizationHeader,
            "Accept" to "application/json",
            "User-Agent" to "triage-companion",
        )

    private val authorizationHeader: String
        get() {
            if (apiKind == JiraAPIKind.DATA_CENTER) {
                return "Bearer $apiToken"
            }

            val encoded = Base64.encodeToString("$email:$apiToken".toByteArray(Charsets.UTF_8), Base64.NO_WRAP)
            return "Basic $encoded"
        }

    fun routeURL(path: String): String {
        val apiBase = if (cloudID != null) {
            "https://api.atlassian.com/ex/jira/${encodedPathComponent(cloudID)}"
        } else {
            baseURL
        }

        return "$apiBase$path"
    }
}
